tutorial/chapter1.md

```diff
@@ -0,0 +1,96 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <thread>
+
+#include <klepsydra/core/event_emitter_middleware_provider.h>
+#include <klepsydra/mem_vision_ocv/image_multiplexer_vision_ocv_provider.h>
+
+#include <klepsydra/mem_core/mem_env.h>
+
+#include <image_processing/image_processor_service.h>
+
+#include "server_user_input.h"
+
+#include "simple_write_service.h"
+
+#include "config.h"
+
+int main(int argc, char* argv[]) {
+    std::string containerName = argc > 1 ? argv[1] : "sample1";
+    std::string folderName = argc > 4 ? argv[4] : TEST_DATA;
+    kpsr::mem::MemEnv environment;
+
+    kpsr::vision_ocv::high_performance::ImageMultiplexerVisionMiddlewareProvider<128> provider(nullptr, "image_processing_feed",
+                                                                                              1024, 640, 16, "body");
+
+    kpsr::EventEmitterMiddlewareProvider<kpsr_tutorial::ResultData> eventEmitter(nullptr,
+                                                                        "image_processing_results",
+                                                                        0, nullptr, nullptr);
+
+    SimpleWriteService simpleWriteService(&environment,
+                                          provider.underlying->getPublisher(),
+                                          folderName,
+                                          true);
+    kpsr_tutorial::ImageProcessingService imageProcessingService(&environment, provider.underlying->getSubscriber(), eventEmitter.getPublisher());
+
+    imageProcessingService.startup();
+    simpleWriteService.startup();
+
+    ServerUserInput serverUserInput;
+
+    std::thread imageProcessingServiceThread = std::thread([&]() {
+        while (true) {
+            simpleWriteService.runOnce();
+            imageProcessingService.runOnce();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            if (!serverUserInput._isRunning) {
+                break;
+            }
+        }
+    });
+    serverUserInput.run();
+
+    imageProcessingServiceThread.join();
+
+    simpleWriteService.shutdown();
+    imageProcessingService.shutdown();
+}
+
+```
+In this example there are some features of Klepsydra that will be introduced and others that have been described in the [kpsr-tutorial - chapter 1](https://github.com/klepsydra-technologies/kpsr-tutorial/blob/master/tutorials/chapter1.md).
+
+* The ```kpsr::vision_ocv::ImageMultiplexerVisionMiddlewareProvider``` is introduced. This provider is a class from the Klepsydra Vision Ocv API and it has two params:
+   * ```kpsr::high_performance::DataMultiplexerMiddlewareProvider```: it is a single-producer, multiple-consumer introduced in the [kpsr-tutorial - chapter 1](https://github.com/klepsydra-technologies/kpsr-tutorial/blob/master/tutorials/chapter1.md)
+   * ```kpsr::vision_ocv::ImageDataFactory```: it is a class from the Klepsydra Vision Ocv API. It is used to handle the allocation of images, copy and resizing
```

Comment:
separate the classes from mem_mdlw and vision_ocv and explain them.

Reply:
Done